chef/cookbooks/provisioner/recipes/bootdisk.rb

```diff
@@ -31,34 +31,35 @@
     basedir = "/dev/disk/by-path"
     dev = nil
     disk_by_path = nil
-    ["-usb-", nil].each do |deviceignore|
-      ::Dir.entries(basedir).sort.each do |path|
-        # Not a symlink?  Not interested.
-        next unless File.symlink?(File.join(basedir, path))
-        # Symlink does not point at a disk?  Also not interested.
-        dev = File.readlink("#{basedir}/#{path}").split("/")[-1]
-        # Prefer devices in a specific order
-        next if (deviceignore and path.include?(deviceignore))
-        disk_by_path = "disk/by-path/#{path}"
-        break if dev =~ /^[hsv]d[a-z]+$/
-        # pci-0000:0b:08.0-cciss-disk0 -> ../../cciss/c0d0
-        break if dev =~ /^c[0-9]+d[0-9]+$/
-        # xen-vbd-51712-part1 -> ../../xvda1
-        break if dev =~ /^xvd[a-z]+$/
-        dev = nil
-        disk_by_path = nil
+    if File.directory?(basedir)
+      ["-usb-", nil].each do |deviceignore|
+        ::Dir.entries(basedir).sort.each do |path|
+          # Not a symlink?  Not interested.
+          next unless File.symlink?(File.join(basedir, path))
+          # Symlink does not point at a disk?  Also not interested.
+          dev = File.readlink("#{basedir}/#{path}").split("/")[-1]
+          # Prefer devices in a specific order
+          next if deviceignore && path.include?(deviceignore)
+          disk_by_path = "disk/by-path/#{path}"
+          break if dev =~ /^[hsv]d[a-z]+$/
+          # pci-0000:0b:08.0-cciss-disk0 -> ../../cciss/c0d0
+          break if dev =~ /^c[0-9]+d[0-9]+$/
+          # xen-vbd-51712-part1 -> ../../xvda1
+          break if dev =~ /^xvd[a-z]+$/
+          dev = nil
+          disk_by_path = nil
+        end
+        break if dev
       end
-      break if dev
     end
-    raise "Cannot find a hard disk!" unless dev
     node[:crowbar_wall][:boot_device] = disk_by_path
+
     # Turn the found device into its corresponding /dev/disk/by-id link.
     # This name should be more stable than the /dev/disk/by-path one.
-
     basedir = "/dev/disk/by-id"
     # /dev/disk/by-id is unstable under VirtualBox so don't rely on it.
     hardware = node[:dmi][:system][:product_name] rescue "unknown"
-    if hardware !~ /VirtualBox/i && File.exists?(basedir)
+    if hardware !~ /VirtualBox/i && File.directory?(basedir)
       bootdisks = ::Dir.entries(basedir).sort.select do |m|
         f = File.join(basedir, m)
         File.symlink?(f) && (File.readlink(f).split("/")[-1] == dev)
```

Comment:
So really, my issue is that we're comparing stuff with `dev` here and now it can be `nil`. Which means this entire block is useless in the `dev.nil?` case and should be in the `else` part I mentioned earlier.